### PR TITLE
Move to a semaphoreslim in the workspace.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Pulled out from https://github.com/dotnet/roslyn/pull/51138 for independent review.

### CODE REVIEW

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs [hunk 1]
```diff
@@ -199,7 +199,7 @@ public async Task InitializeUIAffinitizedServicesAsync(IAsyncServiceProvider asy
             var openFileTracker = await OpenFileTracker.CreateAsync(this, asyncServiceProvider).ConfigureAwait(true);
 
             // Update our fields first, so any asynchronous work that needs to use these is able to see the service.
-            lock (_gate)
+            using (_gate.DisposableWait())
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs:202]: Why not `DisposableWaitAsync`?

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs:202]: was initially making just a straight translation.  but we can def update tehse calls to be DisposableWaitAsync as well .

**Review by sharwell — APPROVED** [review]
How confident are we that reentrancy is impossible?

**Comment by CyrusNajmabadi** [comment]
>How confident are we that reentrancy is impossible?

Not at all confident.  I've already found one case of reentrancy.  Though, IMO, these cases are boken as well.   In cases of non-virtual calls, it's clear what the right code is.  In the case of virtuals, it's not at all clear given that we might have subclasses doing strange things.  So definitely concerning.

**Comment by CyrusNajmabadi** [comment]
CLosing out for now.  I don't feel this can be done particularly safely.  We have reentrancy, and there is a lot of virtual calls here that make it difficult to ensure the same behavior and now risk of deadlocks.

**Pull request closed without merging**